src/kpro_connection.erl

```diff
@@ -0,0 +1,598 @@
+do_init(State0, Sock, Host, Config) ->
+  #state{ parent = Parent
+        , client_id = ClientId
+        } = State0,
+  Debug = sys:debug_options(maps:get(debug, Config, [])),
+  Timeout = get_connect_timeout(Config),
+  %% adjusting buffer size as per recommendation at
+  %% http://erlang.org/doc/man/inet.html#setopts-2
+  %% idea is from github.com/epgsql/epgsql
+  {ok, [{recbuf, RecBufSize}, {sndbuf, SndBufSize}]} =
+    inet:getopts(Sock, [recbuf, sndbuf]),
+    ok = inet:setopts(Sock, [{buffer, max(RecBufSize, SndBufSize)}]),
+  SslOpts = maps:get(ssl, Config, false),
+  Mod = get_tcp_mod(SslOpts),
+  NewSock = maybe_upgrade_to_ssl(Sock, Mod, SslOpts, Timeout),
+  %% from now on, it's all packet-4 messages
+  ok = setopts(NewSock, Mod, [{packet, 4}]),
+  ok = sasl_auth(Host, NewSock, Mod, ClientId, Timeout, get_sasl_opt(Config)),
+  Versions =
+    case Config of
+      #{query_api_versions := false} -> ?undef;
+      _ -> query_api_versions(NewSock, Mod, ClientId, Timeout)
+    end,
+  State = State0#state{mod = Mod, sock = NewSock, api_vsns = Versions},
+  proc_lib:init_ack(Parent, {ok, self()}),
+  ReqTimeout = get_request_timeout(Config),
+  ok = send_assert_max_req_age(self(), ReqTimeout),
+  Requests = kpro_sent_reqs:new(),
+  %% from now on, enter `{active, once}' mode
+  %% NOTE: ssl doesn't support `{active, N}'
+  ok = setopts(NewSock, Mod, [{active, once}]),
+  loop(State#state{requests = Requests, req_timeout = ReqTimeout}, Debug).
+
+query_api_versions(Sock, Mod, ClientId, Timeout) ->
+  Req = kpro_req_lib:make(api_versions, 0, []),
+  Rsp = inactive_request_sync(Req, Sock, Mod, ClientId, ?API_VERSIONS_CORRID,
+                              Timeout, failed_to_query_api_versions),
+  #kpro_rsp{api = api_versions, vsn = 0, msg = Body} = Rsp,
+  ErrorCode = find(error_code, Body),
+  case ErrorCode =:= ?kpro_no_error of
+    true ->
+      Versions = find(api_versions, Body),
+      F = fun(V, Acc) ->
+          API = find(api_key, V),
+          MinVsn = find(min_version, V),
+          MaxVsn = find(max_version, V),
+          case is_atom(API) of
+            true ->
+              %% known API for client
+              Acc#{API => {MinVsn, MaxVsn}};
+            false ->
+              %% a broker-only (ClusterAction) API
+              Acc
+          end
+      end,
+      lists:foldl(F, #{}, Versions);
+    false ->
+      erlang:error({failed_to_query_api_versions, ErrorCode})
+  end.
+
+% Send request to active = false socket, and wait for response.
+inactive_request_sync(#kpro_req{api = API, vsn = Vsn} = Req,
+                      Sock, Mod, ClientId, CorrId, Timeout, ErrorTag) ->
+  ReqBin = kpro_req_lib:encode(ClientId, CorrId, Req),
+  ok = setopts(Sock, Mod, [{active, false}]),
+  try
+    ok = Mod:send(Sock, ReqBin),
+    {ok, RspBin} = Mod:recv(Sock, _Len = 0, Timeout),
+    {CorrId, Body} = decode_corr_id(RspBin),
+    kpro_rsp_lib:decode(API, Vsn, Body, _IgnoredRef = make_ref())
```

Comment:
Could you pass a dummy value instead of creating a new and apparently unused ref here?